Add best practice - Use the gradle.properties File (#33556)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/best-practices/best_practices_general.adoc

=== Tags

<<tags_reference.adoc#tag:upgrades,`#upgrades`>>

[[use_the_gradle_properties_file]]
== Set build flags in `gradle.properties`

> :keywords: properties,structuring-builds

Reviewer: I am temporarily removing keywords due to the new search algorithm I just deployed. I will add them back if appropriate:
```suggestion
```

Set Gradle build property flags in the `gradle.properties` file.

=== Explanation

Instead of using command-line options or environment variables, set build flags in the root project's `gradle.properties` file.

Gradle comes with a long list of <<build_environment.adoc#sec:gradle_configuration_properties,Gradle properties>>, which have names that begin with `org.gradle` and can be used to configure the behavior of the build tool.
You should familiarize yourself with these properties, which can have **enormous** impacts on your build's performance.

You should not rely on supplying these properties on the CLI with every Gradle invocation.
Providing these properties on the command line is meant for temporary testing and debugging, and is easy to forget across different environments.

> A permanent, idiomatic location to set and share these properties across different machines and between developers is in the `gradle.properties` file located in the root project directory.

Reviewer: This is too much in one sentence, let's shorten it:
```suggestion
A permanent, idiomatic location to set and share these properties is in the `gradle.properties` file located in the root project directory.
```
Maybe a second sentence here about sharing the file and checking it into a repo?

> You should be aware of the default values of properties your build uses and avoid explicitly setting properties to their default values.

Reviewer: ```suggestion
You should understand the default values of the properties your build uses and avoid explicitly setting properties to those defaults.
```

> If Gradle ever changes the default value of a property, this will not be a silent change, but will involve a standard <<feature_lifecycle.adoc#backwards_compatibility,deprecation cycle>>.

Reviewer: ```suggestion
Any change to a property's default value in Gradle will follow the standard <<feature_lifecycle.adoc#backwards_compatibility,deprecation cycle>>, and users will be properly notified.
```

NOTE: Properties set this way are not inherited across build boundaries when using <<composite_builds.adoc#composite_builds,composite builds>>.

=== Example

==== Don't Do This

[.multi-language-sample]
=====
[source, kotlin]
----
├── build.gradle.kts
└── settings.gradle.kts
----
=====
[.multi-language-sample]
=====
[source, groovy]
----
├── build.gradle
└── settings.gradle
----
=====

====
include::sample[dir="snippets/bestPractices/useTheGradlePropertiesFile-avoid/kotlin",files="build.gradle.kts[tags=avoid-this]"]
include::sample[dir="snippets/bestPractices/useTheGradlePropertiesFile-avoid/groovy",files="build.gradle[tags=avoid-this]"]
====

This build is run with `gradle run -Dorg.gradle.continue=true`, so that the failure of the `first` task does **not** prevent the `second` task from executing.

This relies on person running the build to remember to set this property, which is error prone and not portable across different machines and environments.

==== Do This Instead

[.multi-language-sample]
=====
[source, kotlin]
----
├── build.gradle.kts
└── gradle.properties
└── settings.gradle.kts
----
=====
[.multi-language-sample]
=====
[source, groovy]
----

> ├── build.gradle.kts

Reviewer: These need to be updated so they are without the `.kts`.

└── gradle.properties
└── settings.gradle.kts
----
=====

====

> include::sample[dir="snippets/bestPractices/useTheGradlePropertiesFile-do/groovy",files="gradle.properties[tags=do-this]"]

Reviewer: Is this missing the kotlin example?

Author: I didn't want a toggle, since the file is exactly the same in either language.  Figured out the formatting.
